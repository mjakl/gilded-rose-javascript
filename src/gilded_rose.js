export class Item {
  constructor(name, sellIn, quality){
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

class TicketStrategy{
  check(item){
    if (isBelowMaxQuality(item)) {
      item.quality = item.quality + 1;
      if (isTicket(item)) {
        if (isLessThan11day(item)) {
          if (isBelowMaxQuality(item)) {
            item.quality = item.quality + 1;
          }
        }
        if (isLessThan6day(item)) {
          if (isBelowMaxQuality(item)) {
            item.quality = item.quality + 1;
          }
        }
      }
    }

    item.sellIn = item.sellIn - 1;
    if (isLessThan0day(item)) {
      item.quality = item.quality - item.quality;
    }
    return item;
  }
}

class FallbackStrategy{
  check(item){

      item.sellIn = item.sellIn - 1;

      if (isLessThan0day(item)) {
        if (isAboveMinQuality(item)) {
          item.quality =item.quality - 1;
        }
      }
    return item;
}
}

class LegendaryStrategy
{
  check(item){
    return item;
  }
}


class CheeseStrategy
{
  check(item){

    if (isBelowMaxQuality(item)) {
      item.quality = item.quality + 1;
    }

    item.sellIn =item.sellIn - 1;

    if (isLessThan0day(item)) {
      if (isBelowMaxQuality(item)) {
        item.quality = item.quality + 1;
      }
    }
    return item;
  }
}

const isEqualToName  = (item, cond) => item.name == cond;

const isLegendary = (item)=> isEqualToName(item, 'Sulfuras, Hand of Ragnaros');
const isNotLegendary = (item)=> !isEqualToName(item, 'Sulfuras, Hand of Ragnaros');

const isTicket = (item)=> isEqualToName(item, 'Backstage passes to a TAFKAL80ETC concert');
const isNotTicket = (item)=> !isEqualToName(item, 'Backstage passes to a TAFKAL80ETC concert');

const isCheese = (item)=> isEqualToName(item, 'Aged Brie');
const isNotCheese = (item)=> !isEqualToName(item, 'Aged Brie');

const isBelowMaxQuality = (item) => item.quality < 50 ;
const isAboveMinQuality = (item) => item.quality > 0 ;

const isLessThan11day = (item) => item.sellIn <11;
const isLessThan6day = (item) => item.sellIn <6;
const isLessThan0day = (item) => item.sellIn <0;

const isAboveMinQualityAndIsNotLegendary = (item) => isAboveMinQuality(item) && isNotLegendary(item)
const isNotCheeseAndIsNotTicket = (item)=> isNotCheese(item) && isNotTicket(item);

const getItem = (items,i)=> items[i];

export class Shop {
  constructor(items=[]){
    this.items = items;
    this.ticketStrategy = new TicketStrategy();
    this.legendaryStrategy = new LegendaryStrategy();
    this.cheesStrategy = new CheeseStrategy();
    this.fallbackStrategy = new FallbackStrategy();

  }
  updateQuality() {
    for (var i = 0; i < this.items.length; i++) {
      let item = getItem(this.items,i);
      if(isTicket(item))
      {
        this.ticketStrategy.check(item);
        continue;
      }
      if(isLegendary(item))
      {
        this.legendaryStrategy.check(item);
        continue;
      }

      if(isCheese(item))
      {
        this.cheesStrategy.check(item);
        continue;
      }

      if (isAboveMinQuality(item)) {
        this.items[i].quality = this.items[i].quality - 1;
      }

      this.fallbackStrategy.check(item);
    }

    return this.items;
  }
}
