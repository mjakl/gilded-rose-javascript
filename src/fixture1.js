import { Item, Shop } from './gilded_rose';

const items = [];

items.push(new Item('+5 Dexterity Vest', 10, 20));
items.push(new Item('Aged Brie', 2, 0));
items.push(new Item('Elixir of the Mongoose', 5, 7));
items.push(new Item('Sulfuras, Hand of Ragnaros', 0, 80));
items.push(new Item('Sulfuras, Hand of Ragnaros', -1, 80));
items.push(new Item('Backstage passes to a TAFKAL80ETC concert', 15, 20));
items.push(new Item('Backstage passes to a TAFKAL80ETC concert', 10, 49));
items.push(new Item('Backstage passes to a TAFKAL80ETC concert', 5, 49));
// this conjured item does not work properly yet
items.push(new Item('Conjured Mana Cake', 3, 6));

items.push(new Item('new item 1', 1, 49));
items.push(new Item('Backstage passes to a TAFKAL80ETC concert', 1, 20));
items.push(new Item('Aged Brie', -2, 20));





const gildedRose = new Shop(items);

export const dayzero = gildedRose.items.map(x=> ({name:x.name, sellIn:x.sellIn, quality:x.quality}));

gildedRose.updateQuality();

export const dayone = gildedRose.items.map(x=> ({name:x.name, sellIn:x.sellIn, quality:x.quality}));

gildedRose.updateQuality();

export const daytwo = gildedRose.items.map(x=> ({name:x.name, sellIn:x.sellIn, quality:x.quality}));

//export default ({daytwo:gildedRose.items, dayone:dayone });
