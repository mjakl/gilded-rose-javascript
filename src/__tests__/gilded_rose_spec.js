import { Shop, Item } from '../gilded_rose';
import {dayzero,dayone,daytwo} from '../fixture1';

describe("Gilded Rose", function () {

    xit("should foo", function () {
        const gildedRose = new Shop([new Item("foo", 0, 0)]);
        const items = gildedRose.updateQuality();
        expect(items[0].name).toEqual("fixme");
    });

    it("should works correctly", function () {
      expect(dayzero).toMatchSnapshot();
      expect(dayone).toMatchSnapshot();
      expect(daytwo).toMatchSnapshot();
    });
});
