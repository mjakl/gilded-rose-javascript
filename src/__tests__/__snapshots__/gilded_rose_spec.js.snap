// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gilded Rose should works correctly 1`] = `
Array [
  Object {
    "name": "+5 Dexterity Vest",
    "quality": 20,
    "sellIn": 10,
  },
  Object {
    "name": "Aged Brie",
    "quality": 0,
    "sellIn": 2,
  },
  Object {
    "name": "Elixir of the Mongoose",
    "quality": 7,
    "sellIn": 5,
  },
  Object {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Object {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 20,
    "sellIn": 15,
  },
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 49,
    "sellIn": 10,
  },
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 49,
    "sellIn": 5,
  },
  Object {
    "name": "Conjured Mana Cake",
    "quality": 6,
    "sellIn": 3,
  },
  Object {
    "name": "new item 1",
    "quality": 49,
    "sellIn": 1,
  },
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 20,
    "sellIn": 1,
  },
  Object {
    "name": "Aged Brie",
    "quality": 20,
    "sellIn": -2,
  },
]
`;

exports[`Gilded Rose should works correctly 2`] = `
Array [
  Object {
    "name": "+5 Dexterity Vest",
    "quality": 19,
    "sellIn": 9,
  },
  Object {
    "name": "Aged Brie",
    "quality": 1,
    "sellIn": 1,
  },
  Object {
    "name": "Elixir of the Mongoose",
    "quality": 6,
    "sellIn": 4,
  },
  Object {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Object {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 21,
    "sellIn": 14,
  },
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 9,
  },
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 4,
  },
  Object {
    "name": "Conjured Mana Cake",
    "quality": 5,
    "sellIn": 2,
  },
  Object {
    "name": "new item 1",
    "quality": 48,
    "sellIn": 0,
  },
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 23,
    "sellIn": 0,
  },
  Object {
    "name": "Aged Brie",
    "quality": 22,
    "sellIn": -3,
  },
]
`;

exports[`Gilded Rose should works correctly 3`] = `
Array [
  Object {
    "name": "+5 Dexterity Vest",
    "quality": 18,
    "sellIn": 8,
  },
  Object {
    "name": "Aged Brie",
    "quality": 2,
    "sellIn": 0,
  },
  Object {
    "name": "Elixir of the Mongoose",
    "quality": 5,
    "sellIn": 3,
  },
  Object {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Object {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 22,
    "sellIn": 13,
  },
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 8,
  },
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 3,
  },
  Object {
    "name": "Conjured Mana Cake",
    "quality": 4,
    "sellIn": 1,
  },
  Object {
    "name": "new item 1",
    "quality": 46,
    "sellIn": -1,
  },
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -1,
  },
  Object {
    "name": "Aged Brie",
    "quality": 24,
    "sellIn": -4,
  },
]
`;
